bin/pytorch_inference/Main.cc

```diff
@@ -8,100 +8,121 @@
 #include <core/CLogger.h>
 #include <core/CProcessPriority.h>
 #include <core/CRapidJsonLineWriter.h>
-#include <core/CoreTypes.h>
+
+#include <seccomp/CSystemCallFilter.h>
 
 #include <ver/CBuildInfo.h>
 
 #include <api/CIoManager.h>
 
-#include <seccomp/CSystemCallFilter.h>
-
 #include "CBufferedIStreamAdapter.h"
 #include "CCmdLineParser.h"
+#include "CCommandParser.h"
 
 #include <rapidjson/ostreamwrapper.h>
 #include <torch/script.h>
 
-#include <boost/optional.hpp>
-
 #include <memory>
 #include <string>
 
-// For ntohl
-#ifdef Windows
-#include <WinSock2.h>
-#else
-#include <netinet/in.h>
-#endif
+namespace {
+const std::string INFERENCE{"inference"};
+const std::string ERROR{"error"};
+}
 
-using TFloatVec = std::vector<float>;
+torch::Tensor infer(torch::jit::script::Module& module,
+                    ml::torch::CCommandParser::SRequest& request) {
 
-torch::Tensor infer(torch::jit::script::Module& module, TFloatVec& data) {
     torch::Tensor tokensTensor =
-        torch::from_blob(data.data(), {1, static_cast<std::int64_t>(data.size())})
+        torch::from_blob(static_cast<void*>(request.s_Tokens.data()),
+                         {1, static_cast<std::int64_t>(request.s_Tokens.size())},
+                         at::dtype(torch::kInt32))
             .to(torch::kInt64);
+
     std::vector<torch::jit::IValue> inputs;
+    inputs.reserve(1 + request.s_SecondaryArguments.size());
     inputs.push_back(tokensTensor);
-    inputs.push_back(torch::ones({1, static_cast<std::int64_t>(data.size())})); // attention mask
-    inputs.push_back(
-        torch::zeros({1, static_cast<std::int64_t>(data.size())}).to(torch::kInt64)); // token type ids
-    inputs.push_back(torch::arange(static_cast<std::int64_t>(data.size())).to(torch::kInt64)); // position ids
 
-    torch::NoGradGuard noGrad;
-    auto tuple = module.forward(inputs).toTuple();
-    auto predictions = tuple->elements()[0].toTensor();
+    for (auto args : request.s_SecondaryArguments) {
```

Comment:
One more thing...
```suggestion
    for (const auto& args : request.s_SecondaryArguments) {
```

Reply:
👍  reference yes but it can't be const because later there is non-const access to `.data()`

Reply:
Interestingly, this should cause undefined behaviour because the tensor is then referencing memory from the loop copy of request vector. I wonder if this points to a missing test of non-empty secondary arguments.

Reply:
Actually, calling `to()` is probably what saves you since it'll create a copy. This makes me wonder, should we be doing this here? I'd have thought it would be better to write the values into `std::vector<std::uint64_t>` and keep with just having a reference to this memory for the Tensor. It may need some alignment shenanigans to make the most of library optimisations but that should be manageable. Something to investigate in a follow up anyway.